### PR TITLE
Trim unnessasary leading lines when removing usings

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/45866
Fixes: https://github.com/dotnet/roslyn/issues/34201

These issues are actually duplicates.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs [hunk 1]
```diff
@@ -48,6 +47,8 @@ public override SyntaxNode DefaultVisit(SyntaxNode node)
                 out SyntaxTriviaList finalTrivia)
             {
                 var currentUsings = new List<UsingDirectiveSyntax>(usings);
+                var onlyDeletedUsingsBefore = true;
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs:50]: This name doesn't make much sense. It starts off true, but we haven't deleted any usings yet.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs [hunk 2]
```diff
@@ -83,13 +84,29 @@ public override SyntaxNode DefaultVisit(SyntaxNode node)
...
                             else
                             {
                                 finalTrivia = leadingTrivia;
                             }
                         }
                     }
+                    else
+                    {
+                        if (onlyDeletedUsingsBefore)
+                        {
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs:99]: Should combine to an `else if` to reduce at least one level of nesting

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs [hunk 3]
```diff
@@ -83,13 +84,29 @@ public override SyntaxNode DefaultVisit(SyntaxNode node)
...
+                            if (i > 0 && !passedLeadngTrivia)
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs:100]: Can you comment the logic and what this is doing to be clear what the expectations are here. I'm struggling to work out what `passedLeadingTrivia` is for, without digging in to `ShouldPreserveTrivia` and what the rest of this whole class is doing. When does passedLeadingTrivia become true?

**Review by davidwengier — COMMENTED** [review]
Maybe the existing test coverage is good enough, but would be good to confirm there are tests for when not all of the first group is removed, when all of the last group is removed, and perhaps what happens when there are two groups that are separated, but the setting to separate them is off, even if just to confirm the behaviour.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs:50]: I'm struggling with naming here. In my first commit I called it `firstUsingGroup`, but it is not a good name either IMO. This flag is here to determine the first using that doesn't need to be deleted, e.g.:
```csharp
using Namespace1; //deleted
using Namespace2; // deleted
using Namespace3; // no need to delete. 'onlyDeletedUsingsBefore' is true, so perform logic of possible blank lines trimming
using Namespace4; // no need to delete. 'onlyDeletedUsingsBefore' is false, do not apply any additional logic
```
So naming suggestions are welcome!

Comment (davidwengier) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpRemoveUnnecessaryImportsService.Rewriter.cs:50]: What you just wrote sounds fine to me: `firstUsingNotBeingRemoved`.

**Review by davidwengier — APPROVED** [review]
Thanks

**Pull request merged**